File: java/src/main/java/com/vesoft/nebula/client/graph/data/ResultSet.java
/* Copyright (c) 2022 vesoft inc. All rights reserved.
 *
 * This source code is licensed under Apache 2.0 License.
 */

package com.vesoft.nebula.client.graph.data;

import com.vesoft.nebula.BindingTable;
import com.vesoft.nebula.Row;
import com.vesoft.nebula.Value;
import com.vesoft.nebula.graph.ExecutionResponse;
import com.vesoft.nebula.graph.PlanDescription;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Spliterator;
import java.util.function.Consumer;

public class ResultSet {
    private final ExecutionResponse response;
    private final List<String> columnNames = new ArrayList<>();
    private final String decodeType = "utf-8";

    public static class Record implements Iterable<ValueWrapper> {
        private final List<ValueWrapper> colValues = new ArrayList<>();
        private List<String> columnNames = new ArrayList<>();

        public Record(List<String> columnNames, Row row, String decodeType) {
            if (columnNames == null) {
                return;
            }

            if (row == null || row.values == null) {
                return;
            }

            for (Value value : row.values) {
                this.colValues.add(new ValueWrapper(value, decodeType));
            }

            this.columnNames = columnNames;
        }

        @Override
        public Iterator<ValueWrapper> iterator() {
            return this.colValues.iterator();
        }

        @Override
        public void forEach(Consumer<? super ValueWrapper> action) {
            this.colValues.forEach(action);
        }

        @Override
        public Spliterator<ValueWrapper> spliterator() {
            return this.colValues.spliterator();
        }


        @Override
        public String toString() {
            List<String> valueStr = new ArrayList<>();
            for (ValueWrapper v : colValues) {
                valueStr.add(v.toString());
            }
            return String.format("ColumnName: %s, Values: %s",
                    columnNames.toString(), valueStr.toString());
        }

        /**
         * get column value by index
         *
         * @param index the index of the rows
         * @return ValueWrapper
         */
        public ValueWrapper get(int index) {
            if (index >= columnNames.size()) {
                throw new IllegalArgumentException(
                        String.format("Cannot get field because the key '%d' out of range", index));
            }
            return this.colValues.get(index);
        }

        /**
         * get column value by column name
         *
         * @param columnName the columna name
         * @return ValueWrapper
         */
        public ValueWrapper get(String columnName) {
            int index = columnNames.indexOf(columnName);
            if (index == -1) {
                throw new IllegalArgumentException(
                        "Cannot get field because the columnName '"
                                + columnName + "' is not exists");
            }
            return this.colValues.get(index);
        }

        /**
         * get all values
         *
         * @return the list of ValueWrapper
         */
        public List<ValueWrapper> values() {
            return colValues;
        }

        /**
         * get the size of record
         *
         * @return int the size of columns
         */
        public int size() {
            return this.columnNames.size();
        }

        /**
         * if the column name exists
         *
         * @param columnName the column name
         * @return boolean
         */
        public boolean contains(String columnName) {
            return this.columnNames.contains(columnName);
        }

    }

    public ResultSet(ExecutionResponse resp) {
        if (resp == null || resp.executionOutcome == null) {
            throw new RuntimeException("got null object for server's response");
        }
        this.response = resp;

        if (resp.executionOutcome.result != null
                && resp.executionOutcome.result.columnNames != null) {
            // space name's charset is 'utf-8'
            for (byte[] column : resp.executionOutcome.result.columnNames) {
                this.columnNames.add(new String(column));
            }
        }
    }

    /**
     * the execute result is succeeded
     *
     * @return boolean
     */
    public boolean isSucceeded() {
        return "SUCCESS".equals(new String(response.executionOutcome.gqlStatus.status));
    }

    /**
     * the result data is empty
     *
     * @return boolean
     */
    public boolean isEmpty() {
        return response.executionOutcome.result == null
                || response.executionOutcome.result.records.isEmpty();
    }

    /**
     * get status of execute result
     *
     * @return String
     */
    public String getGqlStatus() {
        return new String(response.executionOutcome.gqlStatus.status);
    }


    /**
     * get latency of the query execute time
     *
     * @return int
     */
    public long getLatency() {
        return response.latencyInUs;
    }

    /**
     * get the PalnDesc
     *
     * @return PlanDescription
     */
    public PlanDescription getPlanDesc() {
        return response.executionOutcome.plan_desc;
    }

    /**
     * get column names of the dataset
     *
     * @return the list of result columns
     */
    public List<String> getColumnNames() {
        return columnNames;
    }

    /**
     * get the size of rows
     *
     * @return int
     */
    public int rowsSize() {
        BindingTable result = response.executionOutcome.result;
        if (result == null || result.records == null) {
            return 0;
        }
        return result.records.size();
    }

    /**
     * get row values with the row index
     *
     * @param index the index of the rows
     * @return Record
     */
    public Record rowValues(int index) {
        if (response.executionOutcome.result == null) {
            throw new RuntimeException("Empty data");
        }
        List<Row> records = response.executionOutcome.result.records;
        if (index >= records.size()) {
            throw new ArrayIndexOutOfBoundsException();
        }
        return new Record(columnNames, records.get(index), decodeType);
    }

    /**
     * get col values on the column key
     *
     * @param columnName the column name
     * @return the list of ValueWrapper
     */
    public List<ValueWrapper> colValues(String columnName) {
        if (response.executionOutcome.result == null) {
            throw new RuntimeException("Empty data");
        }
        int index = columnNames.indexOf(columnName);
        if (index < 0) {
            throw new ArrayIndexOutOfBoundsException();
        }
        List<ValueWrapper> values = new ArrayList<>();
        List<Row> records = response.executionOutcome.result.records;
        for (int i = 0; i < records.size(); i++) {
            values.add(new ValueWrapper(records.get(i).values.get(index), decodeType));
        }
        return values;
    }

    /**
     * get all rows, see {@link Record}
     *
     * @return the list of Row
     */
    public List<Record> getRows() {
        if (response.executionOutcome.result == null) {
            return null;
        }
        List<Record> rows = new ArrayList<>();
        if (response.getExecutionOutcome().getResult().getRecords() == null
                || response.getExecutionOutcome().getResult().getRecords().size() == 0) {
            return rows;
        }
        for (Row row : response.getExecutionOutcome().getResult().getRecords()) {
            rows.add(new Record(columnNames, row, "utf-8"));
        }
        return rows;
    }

    @Override
    public String toString() {
        // When error, print the raw data directly
        if (!isSucceeded()) {
            if (response.getExecutionOutcome().getGqlStatus() == null) {
                return null;
            }
            return new String(response.getExecutionOutcome().getGqlStatus().status);
        }
        int i = 0;
        List<String> rowStrs = new ArrayList<>();
        while (i < rowsSize()) {
            List<String> valueStrs = new ArrayList<>();
            for (ValueWrapper value : rowValues(i)) {
                valueStrs.add(value.toString());
            }
            rowStrs.add(String.join(",", valueStrs));
            i++;
        }
        return String.format("ColumnName: %s,\n Rows: %s",
                columnNames.toString(), rowStrs.toString());
    }
}

File: python/nebulagraph_python/result_set.py
from collections.abc import Iterable, Iterator
from threading import Lock
from typing import Any, List, Optional, Union

from .data import ExtraInfo, PlanInfoNode
from .error_code import ErrorCode
from .proto.graph_pb2 import ExecuteResponse, QueryStats
from .result_table import ResultTable
from .value_wrapper import Row, ValueWrapper


class Record(Iterable[ValueWrapper]):
    def __init__(self, column_names: Optional[List[str]], row: Row):
        self.column_names: List[str] = []
        self.col_values: List[ValueWrapper] = []

        if column_names is None:
            return

        if row is None or not row.values:
            return

        for value in row.values:
            self.col_values.append(value)

        self.column_names = column_names

    def __iter__(self) -> Iterator[ValueWrapper]:
        return iter(self.col_values)

    def __str__(self) -> str:
        value_strs = [str(v.cast()) for v in self.col_values]
        return f"ColumnName: {self.column_names}, Values: {value_strs}"

    def get(self, key: Union[int, str]) -> ValueWrapper:
        if isinstance(key, int):
            if key >= len(self.column_names):
                raise ValueError(
                    f"Cannot get field because the key '{key}' out of range",
                )
            return self.col_values[key]
        index = self.column_names.index(key)
        if index == -1:
            raise ValueError(
                f"Cannot get field because the columnName '{key}' is not exists",
            )
        return self.col_values[index]

    def values(self) -> List[ValueWrapper]:
        return self.col_values

    def size(self) -> int:
        return len(self.column_names)

    def contains(self, column_name: str) -> bool:
        return column_name in self.column_names

    def for_each(self, action):
        for value in self.col_values:
            action(value)

    def spliterator(self):
        return self.col_values.__iter__()


class ResultSet:
    def __init__(self, response: ExecuteResponse):
        if response is None:
            raise RuntimeError("got null object for server's response")

        self.response = response
        self.column_names: List[str] = []
        self.empty = False
        self._index = 0
        self._index_lock = Lock()

        if not response.HasField("result") or not response.result.HasField("meta"):
            self.size = 0
            self.result_table = None
            self.empty = True
            return

        self.result_table = ResultTable(response.result)
        self.column_names.extend(self.result_table.get_column_names())
        self.size = self.result_table.get_total_num_records()

        self.empty = self.size == 0

    def is_succeeded(self) -> bool:
        return self.response.status.code == b"00000"

    def is_empty(self) -> bool:
        return self.empty

    def get_error_code(self) -> ErrorCode:
        return ErrorCode.find(self.response.status.code.decode("utf-8"))

    def get_error_message(self) -> str:
        return self.response.status.message.decode("utf-8")

    def get_latency(self) -> int:
        return self.response.summary.elapsed_time.total_server_time_us

    def get_plan_desc(self) -> PlanInfoNode:
        return PlanInfoNode(self.response.summary.plan_info)

    def get_column_names(self) -> List[str]:
        return self.column_names

    def row_size(self) -> int:
        if self.empty:
            return 0
        return self.size

    def has_next(self) -> bool:
        if self.empty:
            return False
        return self._index < self.size

    def next(self) -> Record:
        if not self.has_next():
            raise StopIteration("no more row record data")
        row = self.result_table.next()
        with self._index_lock:
            self._index += 1
        return Record(self.column_names, row)

    def get_extra_info(self) -> ExtraInfo:
        if not self.response.HasField("summary"):
            return ExtraInfo()

        extra_info = ExtraInfo()
        query_stats: QueryStats = self.response.summary.query_stats
        extra_info.affected_nodes = query_stats.num_affected_nodes
        extra_info.affected_edges = query_stats.num_affected_edges
        extra_info.cursor = self.response.cursor.decode("utf-8")
        extra_info.build_time_us = self.response.summary.elapsed_time.build_time_us
        extra_info.optimize_time_us = (
            self.response.summary.elapsed_time.optimize_time_us
        )
        extra_info.serialize_time_us = (
            self.response.summary.elapsed_time.serialize_time_us
        )
        extra_info.total_server_time_us = (
            self.response.summary.elapsed_time.total_server_time_us
        )
        return extra_info

    def __str__(self) -> str:
        if not self.is_succeeded():
            return self.response.status.message.decode("utf-8")
        return f"ColumnName: {self.column_names}, RowSize: {self.row_size()}, Latency: {self.get_latency()}"

    def as_pandas_df(self) -> Any:
        """Convert result set to pandas DataFrame.

        Returns:
            pandas.DataFrame: DataFrame containing the query results

        Raises:
            ImportError: If pandas is not installed

        """
        try:
            import pandas as pd
        except ImportError:
            raise ImportError(
                "pandas is required to use this method. Please install it using 'pip install pandas'",
            )

        if not self.is_succeeded():
            raise RuntimeError(f"Query failed: {self.get_error_message()}")

        if self.is_empty():
            return pd.DataFrame(columns=self.column_names)

        # Reset index to start
        self._index = 0

        # Build list of rows
        rows = []
        while self.has_next():
            record = self.next()
            row = []
            for val in record.values():
                row.append(val.cast_primitive() if val is not None else None)
            rows.append(row)

        return pd.DataFrame(rows, columns=self.column_names)

    def as_ascii_table(
        self,
        style: str = "table",
        width: int = None,
        min_width: int = 8,
        max_width: int = None,
        padding: int = 1,
        collapse_padding: bool = False,
    ) -> str:
        """Print query results in a formatted table or row-by-row format.

        Args:
            style: Output style - either "table" (default) or "rows"
            width: Fixed width for all columns. If None, width will be auto-calculated
            min_width: Minimum width of columns when using table style
            max_width: Maximum width of columns. If None, no maximum is enforced
            padding: Number of spaces around cell contents in table style
            collapse_padding: Reduce padding when cell contents are too wide

        Examples:
            # Print as table (default)
            result.as_ascii_table()

            # Print as rows
            result.as_ascii_table(style="rows")

            # Customize table formatting
            result.as_ascii_table(width=20, max_width=30, padding=2)

        Returns:
            str: Formatted representation of the results, or error message if query failed

        """
        try:
            from io import StringIO

            from rich import box
            from rich.console import Console
            from rich.table import Table
        except ImportError:
            raise ImportError(
                "The 'rich' library is required to use this method. Please install it using 'pip install rich'.",
            )

        console = Console(file=StringIO(), force_terminal=False)

        if not self.is_succeeded():
            console.print(f"[bold red]Error:[/bold red] {self.get_error_message()}")
            return console.file.getvalue()

        if self.is_empty():
            console.print("[yellow]Empty result set[/yellow]")
            return console.file.getvalue()

        # Reset index to start
        self._index = 0
        column_names = self.get_column_names()

        if style == "rows":
            # Row-by-row format
            row_num = 1
            try:
                while self.has_next():
                    record = self.next()
                    console.print(f"\n[bold blue]Row {row_num}[/bold blue]")
                    for col, val in zip(column_names, record.values()):
                        console.print(f"  [cyan]{col}:[/cyan] {val}")
                    row_num += 1
            except StopIteration:
                pass

        else:
            # Table format
            table = Table(
                box=box.DOUBLE_EDGE,
                show_header=True,
                header_style="bold cyan",
                width=width,
                min_width=min_width,
                padding=padding,
                collapse_padding=collapse_padding,
            )

            for header in column_names:
                table.add_column(header, max_width=max_width)

            try:
                while self.has_next():
                    record = self.next()
                    table.add_row(*[str(v.cast()) for v in record.values()])
            except StopIteration:
                pass

            console.print(table)

        # Print summary
        console.print("\n[bold green]Summary[/bold green]")
        console.print(f"├── [green]Rows:[/green] {self.row_size()}")
        console.print(f"└── [blue]Latency:[/blue] {self.get_latency()}μs")

        return console.file.getvalue()

    def print(
        self,
        style: str = "table",
        width: int = None,
        min_width: int = 8,
        max_width: int = None,
        padding: int = 1,
        collapse_padding: bool = False,
    ):
        """Print the results directly to console with rich formatting.

        Args are the same as as_ascii_table().
        """
        from rich.console import Console

        console = Console()
        console.print(
            self.as_ascii_table(
                style,
                width,
                min_width,
                max_width,
                padding,
                collapse_padding,
            ),
        )
